Issue120: Telemetry Application Should Handle Exceptions Better

kb1lqc opened the pull request:
Implemented #120 fixes which overall makes the program much more robust to configuration file errors such as missing files, missing sections, incorrect user input, etc. It also fails gracefully wherever possible and will prevent the program from progressing for fatal errors such as configuration file import errors. Indicating what is wrong so a false sense of operation is not given.

Several API bugs were also fixed as they were found.

Additionally, comments and docstrings were updated. Formatting was not largely considered so this merge does not include PEP8 updates per our contributions.md.

# Changes
## Global
-Update loggingConfig.ini and telemetry.ini config file reads to check for errors and get stuck in infinite loops if an error occurs. Update once per second with error message.

## telemworker()
- Properly detect bad settings or no data with try/except statements as well as type checking for a dictionary indicating an Error as data is sent back in lists.
- Update docstring
- Changed main data `while(1):` to `while True:`

## / app.route()
- Updated docstring
- Try/except statement for request argument gets
- Added local variable initialization section

## /stations app.route()
- Removed unnecessary `.upper()` on `telemetry.py` line 357. Already present on line 333
- Improved callsign specific query
- added nodeid query to correctly implement nodeid
- Use generateStartStopTimes() instead of manually performing function when defined function already exists
- Updated request exception with IOError, removed unused exceptions
- Properly return an HTTP204 when no stations are heard that match query parameters

## generateStartStopTimes()
- Updated docstring
- Updated starttime sensing to detect when a start time was provided without endtime

## initDB()
- Added ConfigParser exception
- Added Sqlite3 exception
- Changed return into True/False on success
- Updated docstring

## createTelemetryList()
- Updated docstring
- Return temp = [] if KeyError occurs

## sqlInsert()
- Added len(telem) check
- Updated docstring
- Return True or False on insertion success
- Cleaned up SQLite3 exception handling
- Add try statements to compartmentalize exceptions

## iso8601ToEpoch()
- Updated docstring

## queryDB():
- added check for timeTuple == None
- Updated queryDb() docstring
- Removed unnecessary upper() from parameter["CALLSIGN"] variable declaration in queryDb as it is specified to be uppercase when input
- Added debug logger `logger.debug(sql)` to queryDb()
- Moved sqlData variable declaration to top of queryDb()
- Added telemetryConfig.Error exception to queryDb() filename read. Logs error and returns empty sqlData list.
- Removed unnecessary Exceptions from queryDb database connection code, uses ProgrammingError and StandardError only. Both exceptions return empty sqlData list and close database connections.
- queryDb() row to sqlData insertion try/except statement only uses StandardError and closes connection followed by returning sqlData.
- Tested raising StandardError and sqlite3.Error exceptions to confirm those errors are caught and responded to correctly.

## main()
- Added exception handling for ConfigParser
- Updated docstring

## queryStationsDb()
- Update docstring
- Fixed callsign parameter SQL comment to remove erroneous sqlend reference
- Updated comment when no callsign is specified
- Fixed "database" comment spelling
- Added try statement with ConfigParser.Error exception for database filename
- Cleaned up sqlite3 connection and execution exceptions, returns sqlData after an exception and closes the database connection when it is open during an exception.
- Updated logger text when no stations are returned from /stations query
- Added else statement for check if callsing and nodeid are `None`. If length is zero indicating no data in queue a HTTP204 is returned.

## /raw app.rout()
- Improved and simplified request argument exceptions
- Simplified argument checking exceptions to use ValueError only when raising an exception
- Added comments about telemetryDict and fixed spelling error "two"
- Spelling error "dumps"

## querystationsdb()
- Added more informative "ProgrammingError" log string to `queryStationsDb()`

kb1lqc committed:
Updated /raw and queryDb() sections of telemetry.py

kb1lqc committed:
Updated stations and raw app routs and related fuctions.

kb1lqc committed:
Updated remainder of Telemetry.py including main() with proper configparser exception handling

kb1lqc committed:
Added try/except statemement to telemetryworker so a bad configuration file callsign-nodeid fails softly

kb1lqc committed:
Updated detection of error and None response from proxy.GET() call. Last commit had non-working code by accident. This seems much more thorough.

kb1lqc committed:
Updated global python code, telemworker(), and root app.route().

kb1lqc committed:
Merge branch 'master' into issue120

Merging with upstream master which should pull in changes from PR #124

kb1lqc committed:
Updated /stations to properly query nodeid and removed unused timetuple update from callsign specific query.

kb1lqc committed:
Updated generateStartStopTimes() to query for starttime only correctly. Updted /stations to use generateStartStopTimes().

kb1lqc committed:
Removed commented out starttime code

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -29,59 +29,99 @@
 # Start logging after importing modules
-logging.config.fileConfig('loggingConfig.ini')
-logger = logging.getLogger('telemetry')
+try:
```
Confirmed check works as intended.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -29,59 +29,99 @@
-# Should have common file for apps...
 telemetryConfig = ConfigParser.RawConfigParser()
-telemetryConfig.read('telemetry.ini')
+telemetryFile = telemetryConfig.read('telemetry.ini')
```
Check works as intended.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -29,59 +29,99 @@
     This function interfaces the Proxy application via its RESTful interface.
     It is a one-way operation as it makes no sense to POST data to proxy for
-    telemetry to a specific unit with this application.
+    telemetry to a specific unit with this application. This function runs
```
Ok.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -29,59 +29,99 @@
-    # Pragmatically create descriptors for each Faraday connected to Proxy
-    count = config.getint("TELEMETRY", "UNITS")
+        # Pragmatically create descriptors for each Faraday connected to Proxy
+        count = config.getint("TELEMETRY", "UNITS")
```
OK. Count is initialized to 0 prior for good measure.

kb1lqd reviewed (changes requested):
Can you break this up into smaller PR's?

kb1lqd commented:
@kb1lqc This is now 17 days old. Lets get a plan to get this through efficiently by optimized review scope/risk, knowledge, and possibly breaking it up as needed.

kb1lqc commented:
@kb1lqd Yeah I think we should simply tag-up and walk through this. Long-term we should honestly rely on automated tests (submitted with PR) and basically double-check for style and a quick run-through of the code. Until we get that setup we'll probably be a tad slow on this. Unfortunately this came before I needed to step away from development for a week or so which resulted in the extended timeline.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -29,59 +29,99 @@
 
     # initialize variables
     stations = {}
+    count = 0
```
Count of units in config file.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -29,59 +29,99 @@
+            #  Error reading in configs so get stuck in infinite loop indicating problem
+            while True:
+                logger.error("ConfigParse.Error: " + str(e))
+                time.sleep(1)
```
Looks good, infinite loop if config error.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -29,59 +29,99 @@
+            while True:
+                logger.error("ConfigParse.Error: " + str(e))
+                time.sleep(1)
+
```
Looks good, just added try and exeption.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -29,59 +29,99 @@
-    # check for data on telemetry port, if True place into deque
-    while(1):
+    #  Check for data on telemetry port with infinite loop.
+    while True:
```
Looks good, PEP8 mostly.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -29,59 +29,99 @@
+
+            else:
+                # Iterate through each packet and unpack into dictionary
+                logger.debug("Proxy data: " + repr(data))
```
If valid data: Get list of HTTP 200 JSON dictionaries of data
If error from proxy: Get single HTTP 404 JSON dictionary item
Should never be None except if never set.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -129,30 +174,23 @@ def dbTelemetry():
-    except KeyError as e:
-        logger.error("KeyError: " + str(e))
+    except IOError as e:
+        logger.error("IOError: " + str(e))
```
IOerror check due to requests.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -129,30 +174,23 @@ def dbTelemetry():
         return json.dumps({"error": message}), 400
 
     # Create tuple of parameters for SQLite3
-    parameters = {}
```
Assumes local function variables are lost after runtime and reset to default nothing.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -161,7 +199,6 @@ def dbTelemetry():
     parameters["TIMESPAN"] = timespan
     parameters["LIMIT"] = limit
 
-    data = []
```
Assumes local variables are lost after function run-time and don't need re-initialization to clear old values.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -181,22 +218,20 @@ def rawTelemetry():
+    :return: JSON formatted string with data or error. HTTP204 if no data
     """
+    #  Initialize local variables
+    stationData = []
```
Needed to initialize?

kb1lqd replied:
Is needed due to potential for no stations to be present prior to len() check later when no variable is present.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -181,22 +218,20 @@ def rawTelemetry():
-    except KeyError as e:
-        logger.error("KeyError: " + str(e))
+    except IOError as e:
+        logger.error("IOError: " + str(e))
```
IOerror check due to requests.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -221,17 +256,19 @@ def rawTelemetry():
             if limit <= 0:
                 message = "Error: Limit '{0}' is invalid".format(limit)
-                return json.dumps({"error": message}), 400
+                raise ValueError(message)
```
Removed redundant error 400 from below.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -257,10 +294,10 @@ def rawTelemetry():
             for key, value in telemetryDicts.iteritems():
                 # Make sure queue actually has data in it
                 if (len(value) > 0):
-                    stationData = []
```
removed a stationdata list init?

kb1lqd replied:
No reason to erase anything from stationdata. OK.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -270,12 +307,16 @@ def rawTelemetry():
                             break
+                            #  Unsure if break will cause multiple locally connected
+                            #  units to not have data
+
```
OK will have to watch this.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -309,7 +350,12 @@ def rawTelemetry():
+        logger.info("No station data is in queue")
+        return '', 204  # HTTP 204 response cannot have message data
+
+    # Completed our query for "/raw", return json.dumps() and HTTP 200
```
OK, looks good.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -333,18 +381,15 @@ def stations():
 
+    except IOError as e:
+        logger.error("IOError: " + str(e))
+        return json.dumps({"error": str(e)}), 400
```
IOerror check due to requests library.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -354,15 +399,15 @@ def stations():
     parameters["STARTTIME"] = startTime
     parameters["ENDTIME"] = endTime
-    parameters["CALLSIGN"] = callsign.upper()
+    parameters["CALLSIGN"] = callsign
```
Removal OK because config get above forces upper already.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -354,15 +399,15 @@ def stations():
     # Check if no stations returned, if not, return HTTP 204
     if len(data) <= 0:
-        logger.info("Station(s) not heard in last %d seconds", timespan)
+        logger.info("No Station have been heard in last %d seconds", timespan)
```
Typo: station should be "stations"

kb1lqc replied:
Fixed

kb1lqd reviewed (changes requested):
Typo.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
+
+    :param error: HTTP Error
+    :return: JSON formatted string with error description
+    """
```
OK.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
+
+    except ConfigParser.Error as e:
+        logger.error("ConfigParse.Error: " + str(e))
+        return False
```
IOerror check for config parser.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
+            logger.error("Sqlite3.Error: " + str(e))
+            return False
+
+    return True
```
Errors gracefully if database issue. OK.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
+
+    except KeyError as e:
+        logger.error("KeyError: " + str(e))
+        temp = []
```
Only added try and specific key error. izz cool.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
+        logger.error("ConfigParse.Error: " + str(e))
+        return False
+
+    # Change dictionary into list with proper order
```
Specific error for config program.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
+
+        except sqlite3.Error as e:
+            logger.error("Sqlite3.Error: " + str(e))
+            return False
```
OK. Avoids no telem causing failure. A safe way to create SQL commands.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
+        try:
+            # Use connection as context manager to rollback automatically if error
+            with conn:
+                conn.execute(sql,telem)
```
OK. Executing prior developed SQL command.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
+        except sqlite3.Error as e:
+            logger.error("Sqlite3.Error: " + str(e))
+            conn.close()
+            return False
```
You should do a roll-back to throw out SQL edits if error occurs to avoid corrupting database.

kb1lqc replied:
Added rollback() after execution of SQL

kb1lqd reviewed (changes requested):
Need to add SQL safe to corrupted data in event of failure.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -522,32 +639,40 @@ def queryDb(parameters):
     # Iterate through resulting data and create a list of dictionaries for JSON
     try:
-        sqlData = []
+        rows = cur.fetchall()
```
First addition to sqldata. This is running in a loop.

kb1lqd replied:
Fetch all rows that match.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
     """
+
+    #Declare local variables
+    sqlData = []
```
No reason to declare at start of function. Remove.

kb1lqc replied:
Removed

kb1lqc replied:
Actually added back in, I realized that sqlData is a convenient way to return an empty list when an error occurs. This simplifies the downstream logic.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -371,134 +416,206 @@ def stations():
+
+    except StandardError as e:
+        logger.error("StandardError: " + str(e))
+        return sqlData
```
sqlData will always be empty, why return? Remove.

kb1lqc replied:
See above comment for line 601

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -522,32 +639,40 @@ def queryDb(parameters):
+
+    except ConfigParser.Error as e:
+        logger.error("ConfigParse.Error: " + str(e))
+        return sqlData
```
sqlData will always be empty, why return? Remove.

kb1lqc replied:
See above comment for line 601

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -522,32 +639,40 @@ def queryDb(parameters):
-        logger.error("IndexError: " + str(e))
-    except KeyError as e:
-        logger.error("KeyError: " + str(e))
+        return sqlData
```
sqlData will always be empty, why return? Remove.

kb1lqc replied:
See above comment for line 601

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -522,32 +639,40 @@ def queryDb(parameters):
+        logger.error("Sqlite3.Error: " + str(e))
+        logger.error(paramTuple)
+        conn.close()
+        return sqlData
```
sqlData will always be empty, why return? Remove.

kb1lqc replied:
See above comment for line 601

kb1lqd reviewed (changes requested):
sqlData variable returned when always empty.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -575,71 +696,87 @@ def queryStationsDb(parameters):
-        startEpoch = endEpoch - int(parameters["TIMESPAN"])
-        timeTuple = (startEpoch, endEpoch)
+    #  Initialize local variables
+    sqlData = []
```
Don't initialize at start of function if not needed.

kb1lqc replied:
See above comment for line 601

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -575,71 +696,87 @@ def queryStationsDb(parameters):
+
+    except ConfigParser.Error as e:
+        logger.error("ConfigParse.Error: " + str(e))
+        return sqlData
```
sqlData will always be empty, why return? Remove.

kb1lqc replied:
See above comment for line 601

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -575,71 +696,87 @@ def queryStationsDb(parameters):
+    except sqlite3.Error as e:
+        logger.error("Sqlite3.error: " + str(e))
+        logger.error(paramTuple)
+        return sqlData
```
sqlData will always be empty, why return? Remove.

kb1lqc replied:
See above comment for line 601

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -575,71 +696,87 @@ def queryStationsDb(parameters):
-    except KeyError as e:
-        logger.error("KeyError: " + str(e))
+        conn.close()
+        return sqlData
```
sqlData will always be empty, why return? Remove.

kb1lqc replied:
See above comment for line 601

kb1lqd reviewed (changes requested):
more sqlData returns as empty.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -575,71 +696,87 @@ def queryStationsDb(parameters):
-    except KeyError as e:
-        logger.error("KeyError: " + str(e))
+        conn.close()
+        return sqlData
```
Another SQLdata empty?

kb1lqc replied:
See above comment for line 601

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -681,24 +825,41 @@ def generateStartStopTimes(parameters):
-    endEpoch = time.mktime(end)
+    except StandardError as e:
+        logger.error("StandardError: " + str(e))
+        return None
```
Thus function looks OK.

kb1lqd commented on Applications/Telemetry/telemetry.py:
```diff
@@ -712,8 +873,14 @@ def main():
+    except ConfigParser.Error as e:
+        while True:
+            logger.error("ConfigParse.Error: " + str(e))
+            time.sleep(1)
```
Added try and except specific to config parser.

kb1lqd commented:
@kb1lqc Ready for your responses/fixes.

kb1lqc committed:
Fixed typo station to stations

Line 410 had a typo where station should have been stations to be
correctly worded.

kb1lqc committed:
Added Rollback()

In db insertion function sqlInsert() a rollback() was added to prevent any
database corruption from ocurring. This was only done after any sql
execution was performed.

kb1lqc committed:
Removed sqlData initialization

sqlData was initialized well before it was needed. Therefore this was
removed.

kb1lqc committed:
Removed Unnecessary sqlData Returns

A bunch of return statements were unnecessary and therefore removed. Also,
it was unnecessary to return at all prior to creating the actual sqlData
initialization just prior to filling it up.

kb1lqc committed:
Placed sqlData initialization and returns back in

Realized they were there to return an empty list when error occurs.

kb1lqc committed:
Updated docstring with return

Added empty statement for docstring return.

kb1lqd commented:
Approving. This PR runs telemetry as intended and I can query for stations as expected. Both local and RF unit received over several minutes. Some functionality left untested such as proper error handling and SQL rollback better handled by automated testing.
